cherrypick-2.0: ts: Fix "dip" at end of downsampled queries

mrtracy opened the pull request:
Due to the curiosities of downsampling, it is possible for a query to
show a persistent "dip" for the most recent datapoint. This occurs for
downsampled datapoints which are incomplete.

For example, when querying the last hour of data, the UI downsamples to
a resolution where each data point represents 30 seconds of time. In
practical usage, this means each returned point is calculated from 3
underlying 10-second resolution samples queried from the database.
However, the most recent datapoint may be only partially complete; that
is, perhaps only 1 or 2 of the datapoints have yet been collected. In
some graphs, especially those which show rates, this can result in the
appearance of a "dip" when graphing the data.

The fix involves two adjustments:

+ Disallow queries in the future. There are no current use cases for
queries in the future, but it was not previously disallowed
specifically.
+ If the end of the query range is near the current time, we "normalize"
the end timestamp to exclude the most recent data period which may be
incomplete.

Resolves #13537

Release note: Fixed a condition where a persistent trailing dip could
appear in graphs for longer time periods.

Commit:
ts: Fix "dip" at end of downsampled queries

Due to the curiosities of downsampling, it is possible for a query to
show a persistent "dip" for the most recent datapoint. This occurs for
downsampled datapoints which are incomplete.

For example, when querying the last hour of data, the UI downsamples to
a resolution where each data point represents 30 seconds of time. In
practical usage, this means each returned point is calculated from 3
underlying 10-second resolution samples queried from the database.
However, the most recent datapoint may be only partially complete; that
is, perhaps only 1 or 2 of the datapoints have yet been collected. In
some graphs, especially those which show rates, this can result in the
appearance of a "dip" when graphing the data.

The fix involves two adjustments:

+ Disallow queries in the future. There are no current use cases for
queries in the future, but it was not previously disallowed
specifically.
+ If the end of the query range is near the current time, we "normalize"
the end timestamp to exclude the most recent data period which may be
incomplete.

Resolves #13537

Release note: Fixed a condition where a persistent trailing dip could
appear in graphs for longer time periods.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23874)
<!-- Reviewable:end -->

Commit:
storage: Fix flaky time series test

Fixes TestTimeSeriesMaintenanceQueueServer, a test in the storage
package to verify the time series maintenance queue.

This became flaky because this test actually tests a data point at the
current time, and this was running into random troubles in combination
with #13537

The fix is to move "now" to five minutes into the past, which tests the
same thing but avoids the special treatment of the very-near-past.

Fixes #23877

Release note: None

mrtracy commented:
Now includes #23881, which fixes the flaky test created by the first commit.
